File: Problems & Solutions/1833. Maximum Ice Cream Bars/06 January 2022/Approach 1 - Sorting (Greedy)/Maximum Ice Cream Bars.js
var maxIceCream = function(costs, coins) {
    costs.sort((num1, num2) => num1 - num2);

    let n = costs.length;
    let icecream = 0;

    while (icecream < n && costs[icecream] <= coins) {
        coins -= costs[icecream];
        icecream += 1;
    }

    return icecream;
};

File: Problems & Solutions/1833. Maximum Ice Cream Bars/06 January 2022/Approach 2 - Counting Sort (Greedy)/Maximum Ice Cream Bars.cpp
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int n = costs.size();
        int m = *max_element(costs.begin(), costs.end());
        int icecreams = 0;

        vector<int> costsFrequency(m + 1);
        for (auto& cost : costs) {
            costsFrequency[cost]++;
        }

        for (int cost = 1; cost <= m; ++cost) {
            if (costsFrequency[cost] == 0) {
                continue;
            }
            if (coins < cost) {
                break;
            }
            
            int count = min(costsFrequency[cost], coins / cost);
            coins -= cost * count;
            icecreams += count;
        }

        return icecreams;
    }
};